### ISSUE TITLE
[Feature] Add support for enqueueing Kotlin functional types

### ISSUE THREAD

**Issue by felipecsl** [jobrunr/jobrunr#77]
Right now calling `jobScheduler.enqueue` (or `scheduleRecurrently`) with a Kotlin lambda fails since the semantics of how they are implemented internally are different from Java lambdas.

I have a fork of jobrunr with work in progress to support this use case but before I keep going down the path of making it a mergeable change, I wanted to run it by you to see if there's any interest on adding support for that.
If so, I can go ahead and continue polishing it and might be able to submit a pull request in the next few days.

Let me know what your thoughts are!

**Comment by rdehuyss** [comment]
That would be great! It's on my to-do list for quite some time but have a lot going on at work.

I'll make sure the PR goes in fast (most PR's get merged within a week).

**Comment by rdehuyss** [comment]
For the brave Kotlin developers who want to try out Kotlin support:

- checkout the 2.0 branch which includes Kotlin support
- run `./gradlew publishToMavenLocal`
- add a dependency to: org.jobrunr:jobrunr-kotlin-support:1.0.0-SNAPSHOT
- create a KtJobScheduler (see the [tests](https://github.com/jobrunr/jobrunr/blob/2.0/jobrunr-kotlin-support/src/test/kotlin/org/jobrunr/scheduling/KtJobSchedulerTest.kt#L44))
- start enqueueing jobs!

**And please provide feedback** in this issue... I cannot release without feedback (both positive or negative) because I myself don't have any relevant Kotlin project/experience where I can test is extensively.

**Comment by felipecsl** [comment]
I'll switch to the 2.0 and report back!

**Comment by felipecsl** [comment]
I'm getting the following exception when trying to schedule a recurring job:
```
Caused by: org.jobrunr.JobRunrException: JobRunr encountered a problematic exception. Please create a bug report (if possible, provide the code to reproduce this and the stacktrace)
	at org.jobrunr.JobRunrException.shouldNotHappenException(JobRunrException.java:43) ~[jobrunr-core.jar:na]
	at org.jobrunr.jobs.details.JobDetailsGeneratorUtils.createObjectViaMethod(JobDetailsGeneratorUtils.java:65) ~[jobrunr-core.jar:na]
	at org.jobrunr.jobs.details.instructions.JobDetailsInstruction.getObject(JobDetailsInstruction.java:52) ~[jobrunr-core.jar:na]
	at org.jobrunr.jobs.details.instructions.JobDetailsInstruction.invokeInstruction(JobDetailsInstruction.java:40) ~[jobrunr-core.jar:na]
	at org.jobrunr.jobs.details.instructions.AbstractJVMInstruction.invokeInstructionAndPushOnStack(AbstractJVMInstruction.java:18) ~[jobrunr-core.jar:na]
	at org.jobrunr.jobs.details.JobDetailsFinderContext.invokeInstructions(JobDetailsFinderContext.java:76) ~[jobrunr-core.jar:na]
	at org.jobrunr.jobs.details.JobDetailsFinderContext.getJobDetails(JobDetailsFinderContext.java:63) ~[jobrunr-core.jar:na]
	at org.jobrunr.jobs.details.JobDetailsAsmGenerator$JobDetailsFinder.getJobDetails(JobDetailsAsmGenerator.java:189) ~[jobrunr-core.jar:na]
	at org.jobrunr.jobs.details.JobDetailsAsmGenerator.toJobDetails(JobDetailsAsmGenerator.java:24) ~[jobrunr-core.jar:na]
	at org.jobrunr.scheduling.JobScheduler.scheduleRecurrently(JobScheduler.java:362) ~[jobrunr-core.jar:na]
	at org.jobrunr.scheduling.JobScheduler.scheduleRecurrently(JobScheduler.java:326) ~[jobrunr-core.jar:na]
	at org.jobrunr.scheduling.JobScheduler.scheduleRecurrently(JobScheduler.java:292) ~[jobrunr-core.jar:na]
	at org.jobrunr.scheduling.KtJobScheduler.scheduleRecurrently(KtJobScheduler.kt:83) ~[jobrunr-kotlin.jar:na]
	at org.jobrunr.scheduling.KtBackgroundJob.scheduleRecurrently(KtBackgroundJob.kt:96) ~[jobrunr-kotlin.jar:na]
	at dog.stocks.jobrunner.job.StocksDogRecurringJob.schedule(StocksDogRecurringJob.kt:7) ~[main/:na]
	at dog.stocks.jobrunner.StocksDogJobScheduler.scheduleJobs(StocksDogJobScheduler.kt:15) ~[main/:na]
	at dog.stocks.jobrunner.JobServerApplication.run(JobServerApplication.kt:22) ~[main/:na]
	at org.springframework.boot.SpringApplication.callRunner(SpringApplication.java:804) ~[spring-boot-2.4.2.jar:2.4.2]
	... 6 common frames omitted
Caused by: java.lang.IllegalAccessException: class org.jobrunr.jobs.details.JobDetailsGeneratorUtils cannot access a member of class dog.stocks.jobrunner.job.StocksDogRecurringJob$schedule$1 with modifiers "public volatile"
	at java.base/jdk.internal.reflect.Reflection.newIllegalAccessException(Reflection.java:376) ~[na:na]
	at java.base/java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:642) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:559) ~[na:na]
	at org.jobrunr.jobs.details.JobDetailsGeneratorUtils.createObjectViaMethod(JobDetailsGeneratorUtils.java:63) ~[jobrunr-core.jar:na]
	... 22 common frames omitted
```

The code looks like this:
```kotlin
interface StocksDogJob {
  fun call()
}

abstract class StocksDogRecurringJob(private val cronSchedule: String) : StocksDogJob {
  fun schedule() {
    KtBackgroundJob.scheduleRecurrently(cronSchedule, ::call)
  }
}
```
I have a few subclasses of `StocksDogRecurringJob` that implement the interface method `call` which defines what code is gonna run when the job executes. For context, this code run properly with the code from #78 (issue started after migrating to `2.0` branch)

**Comment by rdehuyss** [comment]
Hi @felipecsl , thanks for the feedback! I was able to reproduce it easily.

Two remarks:
- method reference does not work yet in Kotlin because of the fact that method references are compiled to 2 nested classes which makes it more difficult to find the job method.
- you use JobRunr in a way I had not thought about but which makes totally sense. I wonder how you expect this will work in an environment with IoC. Let's say you have multiple classes implementing the StocksDogJob interface and you schedule these as jobs.

When executing the job, JobRunr asks the IoC container for an implementation of StocksDogJob but what if there are multiple implementations? When using Spring, it would result in a NoUnique bean exception.

I could store the exact class (so not the interface) but I wonder how the various IoC containers react to that.
See also: https://github.com/jobrunr/jobrunr/issues/7

**Comment by rdehuyss** [comment]
Hi @felipecsl , I pushed a new commit which should allow for inheritance within jobs.

If you write your job as follows, all should be well (no method reference).
```kotlin
abstract class StocksDogRecurringJob(private val cronSchedule: String) : StocksDogJob {
  fun schedule() {
    KtBackgroundJob.scheduleRecurrently(cronSchedule) { call() }
  }
}
```

Looking forward to your feedback!

**Comment by rdehuyss** [comment]
@felipecsl - method references are now also supported.

**Comment by rdehuyss** [comment]
@felipecsl, @joonasc, @guilhermbrsp - has anybody already taken a look at this? If not, what is holding you back - should I make a jar for you guys?

**Comment by GuilhermBrSp** [comment]
@rdehuyss im receiving `java.lang.IllegalArgumentException: Please provide a lambda expression (e.g. BackgroundJob.enqueue(() -> myService.doWork()) instead of an actual implementation.` error when trying to use  `KtJobScheduler.enqueue` method, not sure if it is a problem with some dependency or configuration that I'm missing.

Im using the **jobrunr-kotlin-support** with **jobrunr-spring-boot-starter** and kotlin jvm version **1.4.30** (same as **jobrunr-kotlin-support**)

I have created a sample project for testing the kotlin support both with a java wrapper and now with the **jobrunr-kotlin-support**, the java wrapper approach is working fine. You can check the code here https://github.com/GuilhermBrSp/jobrunr-kotlin-sample

**Comment by JoonasC** [comment]
@rdehuyss Sorry, I'm currently a bit busy with other work. I'll check out the 2.0 version ASAP.

**Comment by rdehuyss** [comment]
Hi @GuilhermBrSp , the reason for that is that you explicitly use JobRunr 1.2.5. 

In `build.gradle.kts`, can you replace:
```java
    implementation("org.jobrunr:jobrunr:1.2.5")
    implementation("org.jobrunr:jobrunr-spring-boot-starter:1.2.5")
    implementation("org.jobrunr:jobrunr-kotlin-support:1.0.0-SNAPSHOT")
```

with:
```java
    implementation("org.jobrunr:jobrunr:1.0.0-SNAPSHOT") // not needed as jobrunr-kotlin-support pulls this in as a transient dependency
    implementation("org.jobrunr:jobrunr-spring-boot-starter:1.0.0-SNAPSHOT")
    implementation("org.jobrunr:jobrunr-kotlin-support:1.0.0-SNAPSHOT")
```

After doing so, all should be well.

And perhaps I should rename `org.jobrunr:jobrunr-kotlin-support` to `org.jobrunr:jobrunr-kotlin`. 

Thanks for your help testing!

**Comment by GuilhermBrSp** [comment]
Thanks, @rdehuyss! this simple job enqueuing is now working fine with the KTScheduler, soon as I finish my office hour I'm going to extensively test the kotlin support with the other features of jobrunr.

**Comment by felipecsl** [comment]
@rdehuyss I'm still getting the same exception reported here https://github.com/jobrunr/jobrunr/issues/77#issuecomment-777956533 with the latest version. Running locally this new test you added for method references https://github.com/jobrunr/jobrunr/blob/2.0/jobrunr-kotlin-support/src/test/kotlin/org/jobrunr/scheduling/KtJobSchedulerTest.kt#L142 fails for me, by the way.
Using this workaround https://github.com/jobrunr/jobrunr/issues/77#issuecomment-778768907 I'm able to enqueue, however the job ID is incorrect (it's using the name of the abstract class instead of the concrete class, eg.: `dog.stocks.jobrunner.job.StocksDogRecurringJob.call()`), so executing the job fails because it's not able to find the concrete implementation

**Comment by rdehuyss** [comment]
Hi @felipecsl , I committed these fixes but forgot to push them... 🤦

Really sorry for this but now everything should have been pushed. Could you try again?

**Comment by JoonasC** [comment]
@rdehuyss Kotlin support is working fine for me, IMO it can be released :+1:.

**Comment by JoonasC** [comment]
@rdehuyss Btw, I would also like to add a JobActivator for koin, should I add it to the kotlin-support plugin?

**Comment by rdehuyss** [comment]
By all means, go ahead!

**Comment by JoonasC** [comment]
@rdehuyss Ok, I adde a PR #103.

**Comment by rdehuyss** [comment]
I hope somewhere by the end of this month, beginning next month. I'm the 2.0 there will be some behind changes for the Java api and I also want to improve the code quality for Kotlin.

I also have a lot of personal stuff going (house almost collapsing due to a water leak from the city).

**Comment by JoonasC** [comment]
Ok, sorry to hear that your house almost collapsed, hopefully you can get it sorted.

**Comment by shavz** [comment]
Hello, love the project! I'm keen to see the kotlin support as well and/or provide help with testing to get it in. Thanks for all your hard work and I hope the house situation gets sorted soon.

**Comment by yranzou** [comment]
Hi @rdehuyss !
Dashboard is not accessible in v2.0 (with Kotlin support): returns http error 404 .

**Comment by rdehuyss** [comment]
@yranzou , I tried reproducing this but all works fine.

Please check this closed issue if it helps you: https://github.com/jobrunr/jobrunr/issues/109

Otherwise, please setup a small github repo that reproduces this.

**Comment by yranzou** [comment]
Right now, after project update, got error on build jobrunr-spring-boot-starter jar:
`../jobrunr/jobrunr-spring-boot-starter/src/main/java/org/jobrunr/autoconfigure/JobRunrAutoConfiguration.java:89: error: cannot find symbol
                .andBasicAuthentication(properties.getDashboard().getUsername(), properties.getDashboard().getPassword());
                ^
  symbol:   method andBasicAuthentication(String,String)
  location: class JobRunrDashboardWebServerConfiguration`

**Comment by rdehuyss** [comment]
Please use the latest master branch. I'm preparing everything to release it soon.

I had a merge conflict in the 2.0 branch where I made a mistake.

**Comment by yranzou** [comment]
> @yranzou , I tried reproducing this but all works fine.
> 
> Please check this closed issue if it helps you: #109
> 
> Otherwise, please setup a small github repo that reproduces this.

@rdehuyss , please check this repo to reproduce dashboard error 404
https://github.com/yranzou/jobrunr-dashboard-kotlin

**Comment by rdehuyss** [comment]
@yranzou , i checked both in chrome and Firefox and the dashboard opens immediately 🤔. Which browser do you use?

**Comment by rdehuyss** [comment]
Solved with the release of JobRunr v2.0.0. 

I hope you guys enjoy using JobRunr with Kotlin!!

**Comment by yranzou** [comment]
> @yranzou , i checked both in chrome and Firefox and the dashboard opens immediately . Which browser do you use?
Chromium, Firefox and Postman. Looks like something was wrong with my local build from sources. Now everything works like a charm with 2.0.0 from Maven repo. Thanks!!!
